api/v1alpha3/vspheremachine_webhook.go

```diff
@@ -25,3 +32,60 @@ func (r *VSphereMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
 		For(r).
 		Complete()
 }
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-vspheremachine,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=vspheremachines,versions=v1alpha3,name=validation.vspheremachine.infrastructure.x-k8s.io,sideEffects=None
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *VSphereMachine) ValidateCreate() error {
+	var allErrs field.ErrorList
+	spec := r.Spec
+
+	if spec.Network.PreferredAPIServerCIDR != "" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "PreferredAPIServerCIDR"), spec.Network.PreferredAPIServerCIDR, "cannot be set, as it will be removed and is no longer used"))
+	}
+
+	if spec.ProviderID != nil {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "providerID"), "cannot be set on creation"))
+	}
+
+	for _, device := range spec.Network.Devices {
+		for _, ip := range device.IPAddrs {
+			if _, _, err := net.ParseCIDR(ip); err != nil {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "network", "devices", "ipAddrs"), ip, "ip addresses should be in the CIDR format"))
```

Comment:
I would use index variables for device and IP, and include the index in the field path (e.g. devices[i].ipAddrs[j]).

Also, I think the error message text can be simplified to "must be in CIDR format".

Reply:
@ncdc - do you mean something like:
```
allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "network", fmt.Sprintf("devices[%v]", i), fmt.Sprintf("ipAddrs[%v]", j)), ip, "must be in CIDR format"))
```

Reply:
Yes except I would use `%d`